crates/jstzd/src/docker/runnable_image.rs

```diff
@@ -0,0 +1,371 @@
+use log::info;
+use std::{
+    collections::{BTreeMap, HashMap, HashSet},
+    net::IpAddr,
+    sync::Arc,
+};
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum AccessMode {
+    ReadOnly,
+    ReadWrite,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum MountType {
+    Bind,
+    Volume,
+    Tmpfs,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Mount {
+    pub access_mode: AccessMode,
+    pub mount_type: MountType,
+    pub source: Option<String>,
+    pub target: Option<String>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Host {
+    Addr(IpAddr),
+    HostGateway,
+}
+
+impl ToString for Host {
+    fn to_string(&self) -> String {
+        match self {
+            Host::Addr(ip) => ip.to_string(),
+            Host::HostGateway => "host-gateway".to_string(),
+        }
+    }
+}
+
+type HostPort = u16;
+type ContainerPort = u16;
+pub struct RunnableImage<I: Image> {
+    pub image: I,
+    pub container_name: String,
+    pub overridden_cmd: Option<Vec<String>>,
+    pub network: Option<String>, //TODO: https://linear.app/tezos/issue/JSTZ-111/support-for-adding-network-to-a-container
+    pub env_vars: BTreeMap<String, String>,
+    pub hosts: BTreeMap<String, Host>,
+    pub mounts: Vec<Mount>, //TODO: https://linear.app/tezos/issue/JSTZ-110/implement-mounts-for-runnable-image
+    pub ports: BTreeMap<HostPort, ContainerPort>,
+}
+
+impl<I: Image> RunnableImage<I> {
+    pub fn new(image: I, container_name: impl Into<String>) -> Self {
+        Self {
+            image,
+            container_name: container_name.into(),
+            overridden_cmd: None,
+            network: None,
+            env_vars: BTreeMap::new(),
+            hosts: BTreeMap::new(),
+            mounts: Vec::new(),
+            ports: BTreeMap::new(),
+        }
+    }
+
+    pub fn with_overridden_cmd(mut self, cmd: Vec<String>) -> Self {
+        self.overridden_cmd = Some(cmd);
+        self
+    }
+
+    pub fn with_network(mut self, network: impl Into<String>) -> Self {
+        self.network = Some(network.into());
+        self
+    }
+
+    pub fn with_env_var(
+        mut self,
+        key: impl Into<String>,
+        value: impl Into<String>,
+    ) -> Self {
+        self.env_vars.insert(key.into(), value.into());
+        self
+    }
+
+    pub fn with_host(mut self, hostname: impl Into<String>, host: Host) -> Self {
+        self.hosts.insert(hostname.into(), host);
+        self
+    }
+
+    pub fn with_mount(mut self, mount: Mount) -> Self {
+        self.mounts.push(mount);
+        self
+    }
+
+    pub fn with_port(
+        mut self,
+        host_port: HostPort,
+        container_port: ContainerPort,
+    ) -> Self {
+        self.ports.insert(host_port, container_port);
+        self
+    }
+
+    pub async fn create_container(&self, client: Arc<Docker>) -> Result<String> {
+        self.all_container_ports_are_exposed()?;
+        let config = Config::<String> {
+            image: Some(self.image.image_name().to_string()),
```

Comment:
`image` should refer to the full uri like `<name>:<tag>` otherwise it will always default tag to latest.
Suggest extending `trait Image` with `fn image_uri() -> String`

Reply:
done!